### PR TITLE
Mark SAN extension critical in SPIFFE CSRs for RFC 5280 and AWS PCA compliance

### PR DESCRIPTION
**Pull request by JamesDowning**
SPIFFE SVIDs use X.509 certificates with an empty `Subject` field, relying entriely on the URI SAN. [RFC 5280 4.2.1.6](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6) requires the SAN extension to be critical when the subject is empty. However Go's [x509.CreateCertificateRequest](https://pkg.go.dev/crypto/x509#CreateCertificateRequest) always marks the SAN extension as non-critical, regardless of whether the subject is populated.

This causes issuance failures with strict certificate authorities such as AWS Private CA, which rejects CSRs that have an empty subject with a non-critical SAN extension.

Using cert-manager's [pki.MarshalSANs](https://pkg.go.dev/github.com/cert-manager/cert-manager@v1.19.4/pkg/util/pki#MarshalSANs) helper we now correctly set the SAN extension to critical when the subject is empty.

I've validated this builds and runs as expected allowing for the issuance of certificates from AWS PCA.

### CODE REVIEW

**Comment by JamesDowning** [comment]
@inteon would it be possible to take another look or get another opinion on this please? 

From https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.6:
```
If subject naming information is present only in the subjectAltName extension (e.g., a key bound only to an email address or URI), then the subject name MUST be an empty sequence and the subjectAltName extension MUST be critical.
```

File: internal/csi/driver/util.go [hunk 1]
```diff
@@ -27,6 +27,8 @@ import (
 	"time"
 
 	"github.com/cert-manager/csi-lib/metadata"
+
+	cmpki "github.com/cert-manager/cert-manager/pkg/util/pki"
```
Comment (wallrj-cyberark) [inline internal/csi/driver/util.go:30]: The failing `pull-cert-manager-csi-driver-spiffe-verify` job is `gci` flagging this import grouping — the new third-party import needs to sit in the same sorted block as `csi-lib/metadata`, not in a separate group:

```go
	cmpki "github.com/cert-manager/cert-manager/pkg/util/pki"
	"github.com/cert-manager/csi-lib/metadata"
```

Running `make fix-golangci-lint` applies this automatically. I reproduced the failure and confirmed the fix locally.

File: internal/csi/driver/util.go [hunk 2]
```diff
@@ -35,7 +37,27 @@ func generatePrivateKey(_ metadata.Metadata) (crypto.PrivateKey, error) {
 }
 
 // signRequest will sign a given X.509 certificate signing request with the given key.
+// When URI SANs are present and the subject is empty (as with SPIFFE SVIDs),
+// RFC 5280 §4.2.1.6 requires the SAN extension to be marked critical.
+// Go's x509.CreateCertificateRequest does not do this, so we marshal the
+// SAN extension ourselves via cert-manager's MarshalSANs helper.
 func signRequest(_ metadata.Metadata, key crypto.PrivateKey, request *x509.CertificateRequest) ([]byte, error) {
+	if len(request.URIs) > 0 {
```
Comment (wallrj-cyberark) [inline internal/csi/driver/util.go]: Minor (non-blocking): the doc comment says "when the subject is empty", but the guard only checks for URIs and `hasSubject` is hard-coded to `false`. For this driver it's harmless — the sole caller in `driver.go` never sets a `Subject`, so it isn't a live bug. If you'd like the helper to honour its own contract (and stay correct should a subject ever be added), gating on an empty subject would do it, and a companion test case with a populated subject asserting `Critical == false` would lock the behaviour in. Entirely optional.

File: internal/csi/driver/driver_test.go [hunk 3]
```diff
@@ -129,3 +134,44 @@ func Test_DriverAnnotationSanitization(t *testing.T) {
 		})
 	}
 }
+
+func Test_signRequest_SAN_critical(t *testing.T) {
```
Comment (wallrj-cyberark) [inline internal/csi/driver/driver_test.go]: Could you add a short doc comment to this test explaining its purpose — i.e. that it guards the RFC 5280 requirement that the SAN extension is marked critical when a SPIFFE CSR has an empty subject? It makes the intent clear to future readers who wonder why the criticality matters here.

**Review by wallrj-cyberark — COMMENTED** [review]
Thanks for this, and for the thorough RFC and AWS PCA references — I've checked the change out and verified it locally (it builds, and the new unit test passes).

The approach is sound: `pki.MarshalSANs(sans, false)` sets `Critical: !hasSubject`, so passing `false` correctly marks the SAN extension critical when the subject is empty.

On whether the CA should be responsible for criticality: that holds for some CAs, but AWS Private CA rejects the CSR at validation time (`MalformedCSRException: CSR must mark the SAN extension critical when it has an empty subject`), so it never reaches the issuance step where it could compensate. RFC 5280 §4.1.2.6 is explicit that the SAN MUST be critical when the subject is empty. The istio ztunnel project hit the identical AWS PCA failure (istio/ztunnel#1431) and its CSR generation marks the SAN extension critical for the same reason (`src/tls/csr.rs`). So this looks correct and low-risk — the only caller (`internal/csi/driver/driver.go`) always produces an empty-subject, single-URI SPIFFE CSR.

The only thing blocking the merge is the failing `verify` job, which is a `gci` import-grouping issue rather than anything in the logic — see the inline comment; `make fix-golangci-lint` resolves it.

*Reviewed with Claude (Opus 4.8)*

File: internal/csi/driver/util.go (hunk 1 above)
Comment (JamesDowning) [reply internal/csi/driver/util.go:30]: Updated, the `cert-manager/pkg/util/pki` import now sits in the same sorted black as `csi-lib/metadata`.

File: internal/csi/driver/util.go (hunk 2 above)
Comment (JamesDowning) [reply internal/csi/driver/util.go]: Decided to honour the contract here, I've gated on an actually empty subject via cert-manager `IsASN1SubjectEmpty` instead of hard-coding `hasSubject=false`. Behaviour should be unchanged in practice as you've pointed out but will behave as written should a subject ever be added.

Also updated `Test_signRequest_SAN_criticality` to test both cases.

File: internal/csi/driver/driver_test.go (hunk 3 above)
Comment (JamesDowning) [reply internal/csi/driver/driver_test.go]: Added a doc comment here.

**Comment by JamesDowning** [comment]
@wallrj-cyberark 

Really appreciate the review thank you, I've addressed all of the comments so let me know next steps when you get a chance. Cheers.

File: internal/csi/driver/util.go [hunk 4]
```diff
@@ -34,8 +36,34 @@ func generatePrivateKey(_ metadata.Metadata) (crypto.PrivateKey, error) {
...
+		for i, u := range request.URIs {
+			uriStrings[i] = u.String()
+		}
+		sans := cmpki.GeneralNames{UniformResourceIdentifiers: uriStrings}
+		// no subject, so MarshalSANs marks the SAN critical
+		sanExt, err := cmpki.MarshalSANs(sans, false)
+		if err != nil {
+			return nil, err
+		}
+		request.ExtraExtensions = append(request.ExtraExtensions, sanExt)
```
Comment (wallrj-cyberark) [inline internal/csi/driver/util.go:62]: Non-blocking observation: I _think_ there may be an edge case if the caller ever passed a `CertificateRequest` that already had a SAN extension in `ExtraExtensions` — the `append` would produce a CSR with duplicate SAN extensions (OID 2.5.29.17), which I believe violates RFC 5280 §4.2. I'm not entirely certain whether the Go stdlib deduplicates within `ExtraExtensions` or not, though.

In practice this is safe today — the sole caller at `driver.go:537` creates a fresh `x509.CertificateRequest` with no `ExtraExtensions` — so this is purely a defensive consideration. Feel free to ignore.

**Review by wallrj-cyberark — COMMENTED** [review]
I have verified the updated code locally: builds clean, both test cases pass, and `golangci-lint` reports no issues. The `IsASN1SubjectEmpty` guard is the right approach — it honours the documented contract and stays correct should a subject ever be added.

**Re the Copilot comments** — both were valid observations on the original code, and both are now addressed:

1. **`util.go` — `hasSubject` hard-coded to `false`**: the original guard only checked `len(request.URIs) > 0` without verifying whether the subject was actually empty. While harmless in practice (the sole caller never sets a `Subject`), the code did not match its documented contract. Gating on `IsASN1SubjectEmpty` addresses this properly.

2. **`driver_test.go` — missing non-empty subject test case**: without a negative case, a regression into unconditionally marking the SAN critical would go undetected. The "non-empty subject leaves SAN non-critical" case closes this gap.

All CI is green. One minor (non-blocking) observation inline.

File: internal/csi/driver/util.go [hunk 5]
```diff
@@ -35,7 +37,27 @@ func generatePrivateKey(_ metadata.Metadata) (crypto.PrivateKey, error) {
...
+		sans := cmpki.GeneralNames{UniformResourceIdentifiers: uriStrings}
+		// hasSubject=false → Critical=true
+		sanExt, err := cmpki.MarshalSANs(sans, false)
+		if err != nil {
+			return nil, err
+		}
+		request.ExtraExtensions = append(request.ExtraExtensions, sanExt)
+		// Clear URIs to avoid a duplicate SAN extension from the stdlib.
+		request.URIs = nil
+	}
```
[the remark this answers is not in the document]
Comment (wallrj-cyberark) [reply internal/csi/driver/util.go]: Valid observation — the contributor addressed this in 4f03f6a by gating on `cmpki.IsASN1SubjectEmpty(asn1Subject)` instead of hard-coding `hasSubject=false`. The SAN extension is now only marked critical when the subject is genuinely empty.

File: internal/csi/driver/driver_test.go [hunk 6]
```diff
@@ -129,3 +134,44 @@ func Test_DriverAnnotationSanitization(t *testing.T) {
...
+			break
+		}
+	}
+	require.NotNil(t, sanExt, "CSR must contain the SAN extension (OID 2.5.29.17)")
+	require.True(t, sanExt.Critical, "SAN extension must be critical when subject is empty")
+
+	// Verify the SPIFFE URI is present in the parsed CSR.
+	require.Len(t, csr.URIs, 1)
+	require.Equal(t, spiffeID, csr.URIs[0].String())
+}
```
[the remark this answers is not in the document]
Comment (wallrj-cyberark) [reply internal/csi/driver/driver_test.go:197]: Agreed — the contributor added a "non-empty subject leaves SAN non-critical" test case in 4f03f6a, which guards against this regression.

**Comment by wallrj-cyberark** [comment]
@JamesDowning This change has been included in [v0.14.0](https://github.com/cert-manager/csi-driver-spiffe/releases/tag/v0.14.0), which is now published.

Installation instructions are at https://cert-manager.io/docs/usage/csi-driver-spiffe/installation/

If you are able to install the new release and verify the fix against AWS PCA in your environment, that would be much appreciated. Thank you for the contribution.
